oak/server/dev/dev_oak_loader.cc

```diff
@@ -0,0 +1,99 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "dev_oak_loader.h"
+
+#include "absl/memory/memory.h"
+#include "asylo/grpc/auth/enclave_server_credentials.h"
+#include "asylo/grpc/auth/null_credentials_options.h"
+#include "asylo/identity/descriptions.h"
+#include "asylo/identity/init.h"
+#include "asylo/util/logging.h"
+#include "asylo/util/statusor.h"
+#include "grpcpp/support/status_code_enum.h"
+#include "include/grpcpp/grpcpp.h"
+
+namespace oak {
+
+DevOakLoader::DevOakLoader() : next_application_id_(0) { InitializeAssertionAuthorities(); }
+
+asylo::StatusOr<oak::CreateApplicationResponse> DevOakLoader::CreateApplication(
+    const oak::ApplicationConfiguration& application_configuration) {
+  std::string application_id = NewApplicationId();
+  LOG(INFO) << "Creating application with ID" << application_id;
+
+  auto runtime = absl::make_unique<OakRuntime>();
+  auto status = runtime->Initialize(application_configuration);
+  if (!status.ok()) {
+    return asylo::Status(status);
+  }
+
+  // Start the runtime.
+  runtime->Start();
+
+  int32_t port = runtime->GetPort();
+  LOG(INFO) << "gRPC server is listening on port: " << port;
+  runtimes_[application_id] = std::move(runtime);
+
+  oak::CreateApplicationResponse response;
+  response.set_application_id(application_id);
+  response.set_grpc_port(port);
+  return response;
+}
+
+grpc::Status DevOakLoader::TerminateApplication(const std::string& application_id) {
+  LOG(INFO) << "Terminating application with ID " << application_id;
+
+  OakRuntime* runtime = runtimes_[application_id].get();
+  if (runtime == nullptr) {
+    LOG(ERROR) << "Unrecognized application ID";
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Unknown application ID");
+  }
+
+  runtime->Stop();
+  LOG(INFO) << "Application with ID " << application_id << " stopped";
+  runtimes_.erase(application_id);
+
+  return grpc::Status::OK;
+}
+
+// Even if we are not running in an enclave, we are still relying on Asylo assertion authorities
+// This allows us to use the same client code to connect to the runtime, and it will potentially
+// allow us to use non-enclave identities in the future.
+void DevOakLoader::InitializeAssertionAuthorities() {
+  LOG(INFO) << "Initializing assertion authorities";
+  asylo::EnclaveAssertionAuthorityConfig null_config;
+  asylo::SetNullAssertionDescription(null_config.mutable_description());
+  std::vector<asylo::EnclaveAssertionAuthorityConfig> configs = {
+      null_config,
+  };
+  asylo::Status status =
+      asylo::InitializeEnclaveAssertionAuthorities(configs.begin(), configs.end());
+  if (!status.ok()) {
+    LOG(QFATAL) << "Could not initialize assertion authorities";
+  }
+  LOG(INFO) << "Assertion authorities initialized";
+}
+
+std::string DevOakLoader::NewApplicationId() {
```

Comment:
I think the point of this loader should be to create a single application. If anything, the Oak Manager or whatever is calling the loader should keep track of how many applications have been created and what they are called. Basically I think the loader should care about a single application, and most of this logic around creating names for the applications, and keeping track of multiple runtimes, should be in the Oak Manager itself.

Perhaps the loader may accept a string representing the application id to help with debugging and logging?

WDYT? Please let me know if you think that's not a good idea, or whether you think it's better to do that later, perhaps after the oak manager is removed.

Reply:
When we will remove an Oak Manager, will we even need application ids?

Reply:
I don't think so

Reply:
Then I'll remove this logic completely as a part of https://github.com/project-oak/oak/issues/431

Reply:
Will be done as a part of https://github.com/project-oak/oak/issues/431